prove_validator.go

```diff
@@ -33,55 +54,114 @@ func (epp *EigenPodProofs) ProveValidatorContainers(oracleBlockHeader *phase0.Be
+	oracleBeaconStateSlot, err := oracleBeaconState.Slot()
+	if err != nil {
+		return nil, err
+	}
+
+	oracleBeaconStateValidators, err := oracleBeaconState.Validators()
+	if err != nil {
+		return nil, err
+	}
+
+	oracleBeaconStateValidatorBalances, err := oracleBeaconState.ValidatorBalances()
+	if err != nil {
+		return nil, err
+	}
+
+	verifyCheckpointProofsCallParams := &VerifyCheckpointProofsCallParams{}
+
+	// Get beacon state top level roots
+	beaconStateTopLevelRoots, err := epp.ComputeBeaconStateTopLevelRoots(oracleBeaconState)
 	if err != nil {
 		return nil, err
 	}
 
-	VerifyValidatorFieldsCallParams.StateRootProof.StateRootProof, err = beacon.ProveStateRootAgainstBlockHeader(oracleBlockHeader)
+	// Get state root proof
+	verifyCheckpointProofsCallParams.ValidatorBalancesRootProof = &ValidatorBalancesRootProof{}
+	stateRootProof, err := beacon.ProveStateRootAgainstBlockHeader(oracleBlockHeader)
 	if err != nil {
 		return nil, err
 	}
 
-	VerifyValidatorFieldsCallParams.OracleTimestamp, err = GetSlotTimestamp(oracleBeaconState, oracleBlockHeader)
+	// prove the validator balances root against the beacon state root
+	balancesRootProof, err := beacon.ProveBeaconTopLevelRootAgainstBeaconState(beaconStateTopLevelRoots, beacon.BALANCES_INDEX)
 	if err != nil {
 		return nil, err
 	}
-	VerifyValidatorFieldsCallParams.ValidatorIndices = make([]uint64, len(validatorIndices))
-	VerifyValidatorFieldsCallParams.ValidatorFieldsProofs = make([]common.Proof, len(validatorIndices))
-	VerifyValidatorFieldsCallParams.ValidatorFields = make([][]Bytes32, len(validatorIndices))
+
+	verifyCheckpointProofsCallParams.ValidatorBalancesRootProof.ValidatorBalancesRoot = *beaconStateTopLevelRoots.BalancesRoot
+	verifyCheckpointProofsCallParams.ValidatorBalancesRootProof.Proof = append(balancesRootProof, stateRootProof...)
+
+	verifyCheckpointProofsCallParams.BalanceProofs = make([]*BalanceProof, len(validatorIndices))
 	for i, validatorIndex := range validatorIndices {
-		VerifyValidatorFieldsCallParams.ValidatorIndices[i] = validatorIndex
-		// prove the validator fields against the beacon state
-		VerifyValidatorFieldsCallParams.ValidatorFieldsProofs[i], err = epp.ProveValidatorAgainstBeaconState(beaconStateTopLevelRoots, oracleBeaconStateSlot, oracleBeaconStateValidators, validatorIndex)
+		balanceRoot, balanceProof, err := epp.proveValidatorBalanceAgainstBeaconState(beaconStateTopLevelRoots, oracleBeaconStateSlot, oracleBeaconStateValidatorBalances, validatorIndex)
 		if err != nil {
 			return nil, err
 		}
 
-		VerifyValidatorFieldsCallParams.ValidatorFields[i] = ConvertValidatorToValidatorFields(oracleBeaconStateValidators[validatorIndex])
+		var pubkeyHash [32]byte
+		pubkeyHashVariable := crypto.Keccak256(oracleBeaconStateValidators[validatorIndex].PublicKey[:])
+		copy(pubkeyHash[:], pubkeyHashVariable)
```

Comment:
Doesn't this also need the 16bytes of 0 padding to 32-byte align?

Reply:
yes!